cpp/src/io/comp/nvcomp_adapter.cpp

```diff
@@ -321,40 +324,68 @@ static void batched_compress_async(compression_type compression,
   CUDF_EXPECTS(nvcomp_status == nvcompStatus_t::nvcompSuccess, "Error in compression");
 }
 
-bool is_aligned(void const* ptr, std::uintptr_t alignment) noexcept
+constexpr bool is_aligned(void const* ptr, std::uintptr_t alignment) noexcept
 {
   return (reinterpret_cast<std::uintptr_t>(ptr) % alignment) == 0;
 }
 
+// since some compressors have extreme temp memory requirements, scale back the number
+// of chunks to process until the request for temp memory succeeds.
+std::pair<rmm::device_buffer, size_t> compress_temp_buffer(compression_type compression,
+                                                           size_t num_chunks,
+                                                           size_t max_uncomp_chunk_size,
+                                                           rmm::cuda_stream_view stream)
+{
+  auto scaled_num_chunks = num_chunks;
+  size_t scale           = 1;
+  while (scaled_num_chunks > 0) {
+    try {
+      auto const temp_size =
+        batched_compress_temp_size(compression, scaled_num_chunks, max_uncomp_chunk_size);
+      rmm::device_buffer buf(temp_size, stream);
+      return std::pair(std::move(buf), scaled_num_chunks);
+    } catch (rmm::bad_alloc& ba) {
+      // don't loop forever...if scaled_num_chunks is already 1, the following divide will
+      // also yield 1
+      if (scaled_num_chunks == 1) { throw ba; }
+      scaled_num_chunks = util::div_rounding_up_safe(num_chunks, ++scale);
+    }
+  }
+  CUDF_FAIL("Cannot allocate temp buffer for compression");
```

Comment:
We should never get to this point. Does the compiler complain without this line?

Reply:
I will try compiling without it with debug on (that seems to have pickier options set).  I would hope it warns about not returning a value (unless the compiler is really smart and can tell that the function either returns or throws `bad_alloc`). I just want some kind of annotation that that line should not be reached.

Reply:
yep, compile fails without that line

Reply:
Alright, a load-bearing unreachable line :)

Reply:
You can, in principle, use [`CUDF_UNREACHABLE`](https://github.com/rapidsai/cudf/blob/branch-23.02/cpp/include/cudf/detail/utilities/assert.cuh#L57) for this use case. In practice we do prefer `CUDF_FAIL` unless this is in device code since the main benefit of the macro is to help the compiler optimize out those code paths and we're unlikely to ever care about that in host code.